nova/virt/vmwareapi/vmops.py

```diff
@@ -2221,6 +2222,68 @@ def manage_image_cache(self, context, instances):
             datastores_info.append((ds, dc_info))
         self._imagecache.update(context, instances, datastores_info)
 
+        if CONF.vmware.image_as_template:
+            self._age_cached_image_templates(dc_info, instances)
+
+    def _age_cached_image_templates(self, dc_info, instances):
+        project_ids = set()
+        for inst in instances:
+            project_ids.add(inst.project_id)
+        for project_id in project_ids:
+            folder_ref = self._get_project_folder(dc_info,
+                                                  project_id=project_id,
+                                                  type_='Images')
+            self._destroy_expired_image_templates(self._session, folder_ref)
+
+    def _destroy_expired_image_templates(self, session, templ_vm_folder_ref):
+        client_factory = session.vim.client.factory
+        task_filter_spec = client_factory.create('ns0:TaskFilterSpec')
+        task_filter_spec.entity = client_factory.create(
+                                        'ns0:TaskFilterSpecByEntity')
+        task_filter_spec.entity.entity = templ_vm_folder_ref
+        task_filter_spec.entity.recursion = "children"
+
+        expired_templ_vms = {}
+        try:
+            task_collector = session._call_method(session.vim,
+                                          "CreateCollectorForTasks",
+                                          session.vim.service_content.taskManager,
+                                          filter=task_filter_spec)
+
+            while True:
+                page_tasks = session._call_method(session.vim,
+                                                  "ReadNextTasks",
+                                                  task_collector,
+                                                  maxCount=10)
+                if len(page_tasks) == 0:
+                    break
+
+                for ti in page_tasks:
```

Comment:
Can we get another break condition here? You seem to iterate over all tasks for all templates in that path - beginning from the oldest to the newest.

I would suggest reversing the order (read `latestPage` first, then use `ReadPreviousTasks`) and saving the last-used time for every template. If that dict has all templates (e.g. `len(last_use_time) == nr_files_in_dir`, would need another query for number of files in the Images dir), we can stop getting tasks. But this would only work, if we want to change belows logic to delete only unused images and would only help in the case all images are in use.

Reply:
Good point. The implementation seems to wrongly consider that the tasks are not in chronological order.

Reply:
Done.